pandas/core/series.py

```diff
@@ -3166,7 +3168,13 @@ def _try_cast(arr, take_fast_path):
         try:
             subarr = maybe_cast_to_datetime(arr, dtype)
             if not is_extension_type(subarr):
-                subarr = np.array(subarr, dtype=dtype, copy=copy)
+                # Take care in creating object arrays (but generators are not
+                # supported, hence the __len__ check):
+                if dtype == 'object' and (hasattr(subarr, '__len__') and
```

Comment:
use is_object_dtype

Reply:
this should be a separate branch rather than a nested if

Reply:
Do you mean

```python
if is_object_dtype(dtype) and (hasattr(subarr, '__len__') and
                          not isinstance(subarr, np.ndarray)):
    [...]
elif not is_extension_type(subarr):
    [....]
```
?